src/libasr/codegen/asr_to_wasm.cpp

```diff
@@ -243,6 +267,203 @@ class ASRToWASMVisitor : public ASR::BaseVisitor<ASRToWASMVisitor> {
     }
 
+    void emit_wasi_imports() {
+        std::vector<ImportFunc> wasi_imports_funcs = {
+            {"proc_exit", {{ASR::ttypeType::Integer, 4}}, {}},
+            {"fd_write", {{ASR::ttypeType::Integer, 4}, {ASR::ttypeType::Integer, 4},
+            {ASR::ttypeType::Integer, 4}, {ASR::ttypeType::Integer, 4} }, {{ASR::ttypeType::Integer, 4}}}
+        };
+
+        for (auto wasi_func:wasi_imports_funcs) {
+            import_function(wasi_func, "wasi_snapshot_preview1");
+        }
+    }
+
+    void emit_if_else(std::function<void()> test_cond, std::function<void()> if_block, std::function<void()> else_block) {
+        test_cond();
+        wasm::emit_b8(m_code_section, m_al, 0x04);  // emit if start
+        wasm::emit_b8(m_code_section, m_al, 0x40);  // empty block type
+        nesting_level++;
+        if_block();
+        wasm::emit_b8(m_code_section, m_al, 0x05);  // starting of else
+        else_block();
+        nesting_level--;
+        wasm::emit_expr_end(m_code_section, m_al);  // emit if end
+    }
+
+    void emit_loop(std::function<void()> test_cond, std::function<void()> loop_block) {
+        uint32_t prev_cur_loop_nesting_level = cur_loop_nesting_level;
+        cur_loop_nesting_level = nesting_level;
+
+        wasm::emit_b8(m_code_section, m_al, 0x03);  // emit loop start
+        wasm::emit_b8(m_code_section, m_al, 0x40);  // empty block type
+
+        nesting_level++;
+
+        emit_if_else(test_cond, [&](){
+            loop_block();
+            wasm::emit_branch(m_code_section, m_al, nesting_level -
+                cur_loop_nesting_level - 1);  // emit_branch and label the loop
+        }, [&](){});
+
+        nesting_level--;
+        wasm::emit_expr_end(m_code_section, m_al);  // end loop
+        cur_loop_nesting_level = prev_cur_loop_nesting_level;
+    }
+
+    void emit_print_int() {
+        uint32_t func_idx = no_of_types;
+        { // type declaration
+            wasm::emit_b8(m_type_section, m_al, 0x60);
+            wasm::emit_u32(m_type_section, m_al, 1); // no of params
+            wasm::emit_b8(m_type_section, m_al, wasm::type::i64);
+            wasm::emit_u32(m_type_section, m_al, 0); // no of results
+            no_of_types++;
+        }
+        /*** Reference Function Prototype ***/
+        wasm::emit_u32(m_func_section, m_al, func_idx);
+
+        /*** Function Body Starts Here ***/
+        uint32_t len_idx_code_section_func_size =
+            wasm::emit_len_placeholder(m_code_section, m_al);
+        wasm::emit_u32(m_code_section, m_al, 4u); // no of local vars
+        for (int i = 0; i < 4; i++) {
+            wasm::emit_u32(m_code_section, m_al, 1u); // count of local vars of this type
+            wasm::emit_b8(m_code_section, m_al, wasm::type::i64);
+        }
+
+        emit_string("-");
+        emit_string("#");
+        uint32_t digits_mem_loc = avail_mem_loc;
+        for (int i = 0; i < 10; i++) {
+            emit_string(std::to_string(i));
+        }
```

Comment:
This is currently inefficient. The `emit_string` has some `i32.store` instructions inside which would be executed everytime the `print_i64` is called. Ideally, they are only needed to be called once when storing string into memory. (The subsequent calls to `print_i64` store the same values at the same locations in memory.)

Ideally, we should extract this out in some other function which we could call once. Since, it seems to work currently, I guess we can fix it in a separate PR.

Reply:
This has been fixed now. The `emit_string` now works at compile-time. It does not use any `i32.store` instructions. The string `location` and `length` information is directly encoded in the form of a `wasm` data segment at compile-time.